File: arrays/best_time_to_buy_and_sell_stock.py
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock/description/

# Time complexity: O(n)
# Space complexity: O(1)
# Trick: Keep track of the minimum price and the maximum profit
from typing import List


def maxProfit(prices: List[int]) -> int:
        min_price = prices[0]
        max_profit = 0
        
        for price in prices[1:]:
            max_profit = max(max_profit, price - min_price)
            min_price = min(min_price, price)
            
        return max_profit

File: arrays/cpp/best_time_to_buy_and_sell_stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int maxProfit = 0;
        int l = 0;
        int r = 0;
        while (r < prices.size()) {
            // potential profit
            if (prices[l] < prices[r]) {
                maxProfit = max(maxProfit, prices[r] - prices[l]);
            } else {
                l = r;
            }
            ++r;
        }
        return maxProfit;
    }
};
